### DISCUSSION TITLE
Dependency Matching Failing Unexpectedly

### DISCUSSION THREAD

**Discussion by mdw00d** [python-poetry/poetry#8952] [Q&A]
poetry is failing to resolve dependencies with the following error:
```
01: 2: fact: no versions of SA match >0.2.1,<0.3.0
02:  2: conflict: no versions of SA match >0.2.1,<0.3.0
03:   2: ! SA (>0.2.1,<0.3.0) is partially satisfied by not SA (==0.2.1)
04:   2: ! which is caused by "SA (0.2.1) depends on ML (>=1.4.4,<2.0.0)"
05:   2: ! thus: SA (>=0.2.1,<0.3.0) requires ML (>=1.4.4,<2.0.0)
06:   2: ! not ML (>=1.4.4,<2.0.0) is satisfied by not ML (>=1.4.4,<2.0.0)
07:   2: ! which is caused by "no versions of ML match >=1.4.4,<2.0.0"
08:   2: ! thus: SA is forbidden
09:   2: ! SA (>=0.2.1,<0.3.0) is satisfied by SA (>=0.2.1,<0.3.0)
10:   2: ! which is caused by "share-fixer depends on SA (^0.2.1)"
11:   2: ! thus: version solving failed
```

The package registry however does contain version 0.2.1 of "SA" referenced above, which has the dependency:
```
ML = {version = "^1.4.4", source = "toolchain"}
```

and the same "toolchain" package registry contains version 1.4.4 of ML (as well as others).   So the assertion on line 5 should be satisfied.  

I do not understand what is being reported on lines 6 and 7, and I do not understand why it reports no matches on line 7.   Somehow it is requiring a version of SA >0.2.1 when it seems the constraint should be SA >=0.2.1.

If someone could shed some light on this, that would be greatly appreciated.   While I have simplified the problem here, we are seeing this pattern with these packages in broader contexts and clearly something is awry somewhere.

If relevant, both the "SA" and "ML" libraries are coming from the same private package registry "toolchain".

**Comment by mdw00d** [comment]
Looking at this further, it appears that the `[package.source]` section is not getting included in the poetry.lock file (and perhaps in whatever in-memory precursor there might be).

A similar problem appear in another project that also leveraged these same two dependencies.  Doing a `poetry remove ML` followed by a `poetry add` specifying the source then caused the source section to be included in the poetry.lock file.   That was my work-around, but I am wondering if there's a bug somewhere causing the `[package.source]` section to not always be generated when it should be?

**Reply by radoering** [reply]
Without a repro we can only guess what's going wrong. It might be related to #8614, which is only fixed in the master branch, or #8928, which is only fixed in the attached PR by now.
